### DISCUSSION TITLE
What's the best way to specify structure layout difference on architectures?

### DISCUSSION THREAD

**Discussion by huoyaoyuan** [dotnet/runtime#55534] [Q&A] [answered]
When trying to modernize P/Invoke in System.Drawing.Common, I saw this:

https://github.com/dotnet/runtime/blob/11c9394f4ef4a1fc66f93b334ee4f29164c4ff44/src/libraries/System.Drawing.Common/src/Interop/Windows/Interop.Comdlg32.cs#L17-L63

Win32 and Win64 uses different packing size for the struct (confirmed in commdlg.h). This is not easy to define in managed code. Defining two types like this looks very error-prone. Does it worth compiling separately for different archs?

**Comment by jkotas** [comment, marked as answer]
You can define two types, but hide the difference behind interface. https://github.com/dotnet/runtime/pull/33899 has real world example of this technique.

**Comment by nguyenhoanghai1605** [comment]
When targeting different architectures like x86 and x64 in P/Invoke scenarios, struct layout can indeed diverge due to alignment and pointer size differences.

While defining two separate types and abstracting via interface (as mentioned in #33899) is a practical approach, here are deeper alternatives and considerations:

## 🧩 1. Use `#if` Preprocessor Directives
If you're building different binaries for different targets, you can conditionally define the struct:

```csharp
#if X64
    [StructLayout(LayoutKind.Sequential)]
    internal struct PRINTDLG
    {
        public int lStructSize;
        public IntPtr hwndOwner;
        // ...
        public short nFromPage;
        public short nToPage;
        public short nMinPage;
    }
#else
    [StructLayout(LayoutKind.Sequential)]
    internal struct PRINTDLG
    {
        public int lStructSize;
        public int hwndOwner;
        // ...
        public short nFromPage;
        public short nToPage;
        public short nMinPage;
    }
#endif

**Reply by huoyaoyuan** [reply]
For future readers: using conditional compilation is strictly non-answer for this problem. The total intent is to express the layout in the same managed assembly.
